src/osd/PrimaryLogPG.cc

```diff
@@ -2394,6 +2404,154 @@ PrimaryLogPG::cache_result_t PrimaryLogPG::maybe_handle_manifest_detail(
   return cache_result_t::NOOP;
 }
 
+struct C_ManifestFlush : public Context {
+  PrimaryLogPGRef pg;
+  hobject_t oid;
+  epoch_t last_peering_reset;
+  ceph_tid_t tid;
+  utime_t start;
+  uint64_t offset;
+  uint64_t last_offset;
+  PrimaryLogPG::FlushOpRef manifest_fop;
+  C_ManifestFlush(PrimaryLogPG *p, hobject_t o, epoch_t lpr)
+    : pg(p), oid(o), last_peering_reset(lpr),
+      tid(0), start(ceph_clock_now())
+  {}
+  void finish(int r) override {
+    if (r == -ECANCELED)
+      return;
+    pg->lock();
+    if (manifest_fop->rval < 0) {
+      pg->unlock();
+      return;
+    }
+    manifest_fop->io_results[offset] = r;
+    for (auto &p : manifest_fop->io_results) {
+      if (p.second < 0) {
+	pg->finish_manifest_flush(oid, tid, r, manifest_fop->obc, last_offset, manifest_fop);
+	manifest_fop->rval = r;
+	pg->unlock();
+	return;
+      }
+    }
+    if (manifest_fop->chunks == manifest_fop->io_results.size()) {
+      if (last_peering_reset == pg->get_last_peering_reset()) {
+	assert(manifest_fop->obc);
+	pg->finish_manifest_flush(oid, tid, r, manifest_fop->obc, last_offset, manifest_fop);
+	pg->osd->logger->tinc(l_osd_tier_flush_lat, ceph_clock_now() - start);
+      }
+    }
+    pg->unlock();
+  }
+};
+
+int PrimaryLogPG::start_manifest_flush(OpRequestRef op, ObjectContextRef obc, bool blocking,
+				       boost::optional<std::function<void()>> &&on_flush)
+{
+  auto p = obc->obs.oi.manifest.chunk_map.begin();
+  FlushOpRef manifest_fop(std::make_shared<FlushOp>());
+  manifest_fop->op = op;
+  manifest_fop->obc = obc;
+  manifest_fop->flushed_version = obc->obs.oi.user_version;
+  manifest_fop->blocking = blocking;
+  manifest_fop->on_flush = std::move(on_flush);
+  int r = do_manifest_flush(op, obc, manifest_fop, p->first, blocking);
+  if (r < 0) {
+    return r;
+  }
+
+  flush_ops[obc->obs.oi.soid] = manifest_fop;
+  return -EINPROGRESS;
+}
+
+int PrimaryLogPG::do_manifest_flush(OpRequestRef op, ObjectContextRef obc, FlushOpRef manifest_fop,
+				    uint64_t start_offset, bool block)
+{
+  struct object_manifest_t &manifest = obc->obs.oi.manifest;
+  hobject_t soid = obc->obs.oi.soid;
+  ceph_tid_t tid;
+  SnapContext snapc;
+  uint64_t max_copy_size = 0, last_offset = 0;
+
+  map<uint64_t, chunk_info_t>::iterator iter = manifest.chunk_map.find(start_offset); 
+  assert(iter != manifest.chunk_map.end());
+  for (;iter != manifest.chunk_map.end(); ++iter) {
+    if (iter->second.flags == chunk_info_t::FLAG_DIRTY) {
+      last_offset = iter->first;
+      max_copy_size += iter->second.length;
+    }
+    if (get_copy_chunk_size() < max_copy_size) {
+      break;
+    }
+  }
+
+  iter = manifest.chunk_map.find(start_offset);
+  for (;iter != manifest.chunk_map.end(); ++iter) {
+    if (iter->second.flags != chunk_info_t::FLAG_DIRTY) {
+      continue;
+    }
+    uint64_t tgt_length = iter->second.length;
+    uint64_t tgt_offset= iter->second.offset;
+    hobject_t tgt_soid = iter->second.oid;
```

Comment:
I'm not understanding this operation.  It seems like there are actually two states?  In one case, we did the chunking and figured out the object name but we just haven't written it back yet.  In teh second case, we did that already, and overwrite it again, so now it has new/different content.  In that case, we don't actually want to write it back to the  *same* object, do we?  (Or is the content-based naming not addressed yet?  Sorry, my memory is fuzzy on what has been done so far and what hasn't!)

Reply:
There is one state. For example,
1. figure out dirty chunks's size (because we don't know actual dirty chunks, 
 this will be used to check the flush completion)
2. send write messages 
3. receive acks 
 3.1. check completion (whether all dirty chunk we sent are completed or not)
 3.2. do remaining jobs

Regarding remaining jobs,
I thought we needed a way to avoid resource starvation. For example, manifest_flush() needs read(storage) ->
write(network) operations. but, if the object size is huge, flush() repeats such operations without yielding resources.
So, I added the following logic. If the object size is larger than the threshold value (for example, 4MB), 
just stop sending data and wait for the previous flush request to complete. This flush() will continue when  
the previous requests (for flush()) are complete.

Reply:
The flush model sounds right.

I'm thinking of this scenario:

1. write object
2. turn it into many chunks
3. chunks are flushed to lower tier
4. part of object is overwritten in base, and some chunks are marked dirty

At this point, the implemenation of flush above is assuming that the chunks are mutable, so just marking them dirty so they can be updated later is sufficient.  This is fine for normal tiering, but won't work for dedup (where the chunk is immutable, and the manifest chunk isn't dirty, it's more like 'dangling').  Maybe that's fine for now, but what is the plan here?  Will these chunks have a state bit like "immutable" or "cas" so that the base tier knows how to deal with them?  Because I think the "mark chunks dirty when we overwrite parts of the object" code will need to change.  This mid-point is a but odd because I don't think you would actually chunk an object into mutable chunks... chunking is only really useful for the dedup case.

Reply:
I think the chunk (both dedup and normal tier) can be "mutable" because we can consider following scenarios.

* Chunk states for normal tiering (no problem):
Missing -> (overwritten) -> dirty -> (flushed) ->clean or missing -> (overwritten) -> dirty
-> (flushed) -> clean or missing...

* Chunk states for dedup tiering (no dangling):
Missing -> (overwritten) -> dirty -> (flushed) -> clean or missing -> (overwritten) -> dirty 
-> (if the chunk was flushed, we can see old chunk,target_oid (== old cas object id). 
So, just send decrease reference message to old cas object and then, flush()) ->
(if flush is finished, chunk.target_oid can be updated by new value) -> clean or missing ...

We only need to check whether the chunk was flushed and send decrease message before the chunks is flushed to lower tier.  If the cas object's reference is 0, this will be deleted.

Did I misunderstand your question or are there any uncertainties in what I said?

Reply:
That makes sense!